### ISSUE TITLE
How to set `chunk_size` in a dataset preparation script

### ISSUE THREAD

**Issue by jwkirchenbauer** [Lightning-AI/lit-gpt#907]
**TL;DR:** How should one set `chunk_size` when they prepare a (pretraining) dataset for training?

@awaelchli might be the person with the most relevant knowledge based on the commits I'm looking at.

---

Hi there, I'm trying to implement a "prepare" script of my own at the intersection of examples like `scripts/prepare_redpajama.py` and `scripts/prepare_slimpajama.py`. I am aware of the fact that the former writes to a `lit_gpt.packed_dataset.PackedDataset` format and the latter writes to a `lightning.data.streaming.dataset.StreamingDataset` format. 

In either case, especially if they are different, I am trying to understand how to set the `chunk_size` argument. The redpajama script has a comment `block size + 1 for causal` that suggests that the `chunk_size` is set as some multiple of the desired sequence length/block size for the model you want to train on this data. The default in the slimpajama script seems to match this convention too `(2049 * 16384)` where 2049=2048+1 and 16384 is a power of 2, perhaps specifically because of something like 16384/64 = 256 maybe for 64 GPUs or something? a [wandb report](https://wandb.ai/awaelchli/tinyllama/reports/TinyLlama-Reproduction--Vmlldzo2MTQwMjc2?accessToken=8mpemnbzi4q34zwew69mxda86610muaqy6vaybfg219msnnei5bnv5noydglepku) suggets this code was developed for a 64 A100 training run idk...

However, looking into how the actual data processing function works, while the `Tokenizer` is instantiated from a checkpoint dir, which means it _could_ have access to the `config.block_size` argument say to truncate/pad sequences or something, the `tokenizer.encode` calls won't use it unless it is manually passed into each call as `max_length` (because it's not set internally by the `Tokenizer`).

Therefore, I am trying to understand what constraints there are on how this `chunk_size` argument is supposed to be set? My rough guess is that we might be trying to write a "chunk" that corresponds to a "batch" at training time. If this is true, is this a breaking constraint, or an optimization? What is the best practice here? Thanks!

(If it's easier to focus on one dataset format, let's say I'm targeting a `PackedDataset`. I can sort of follow how this argument affects the code flow here, following the `StreamingDataset` class' use of chunks tbh, but if Adrian sees this then the rule of thumb for a `StreamingDataset` is fine also 🙂)

**Comment by xyzhang626** [comment]
want to understand, too. I also notice [tinyllama](https://github.com/jzhang38/TinyLlama/blob/main/scripts/prepare_slimpajama.py#L77) use a far small default chunk size `2049 * 1024`.

**Comment by awaelchli** [comment]
Hey @jwkirchenbauer The Packed Dataset is the old implementation and StreamingDataset is our new solution for streaming data in the cloud efficiently. The concept of a chunk is the same in both: Since the dataset is terrabytes of data, we can't save the entire thing in a single file. But we also can't save every sample as it's own individual file, because that would generate too many files (accessing files in the cloud e.g. S3 is expensive). That's why we chunk the data up in sizable pieces. In particular here, we have 2049 * 16384 tokens in one chunk, that's  ~134 Megabytes per chunk. When we run a training with 8 GPUs in a machine, and each GPU uses several dataloader workers, then these all these processes will download 134 MB each. The streaming dataset also prefetches these chunks so the next chunk is ready, and deletes ones that were completely consumed. A good chunk size is one that corresponds to a few hundred MB, so that all the chunks fit well on the disk of the machine. You want to choose the chunk size small enough that they fit into the machines disk and large enough that you don't hit rate limits from the remote filesystem.

**Comment by jwkirchenbauer** [comment]
Okay thank you for the explanation. So in conclusion, there is no fundamental limitations on how you set chunk size, especially if not reading from the cloud, right? Also, prepare-time chunk size need not match train-time chunk size.

**Comment by awaelchli** [comment]
So in conclusion, there is no fundamental limitations on how you set chunk size, especially if not reading from the cloud, right?

For local execution no, it will work fine. But the streaming aspect is designed for cloud, and that's where the chunk size matters. 

> Also, prepare-time chunk size need not match train-time chunk size.
> (with the caveat that for a large chunk size, some unfilled chunks might contain a ton of separator tokens)

After preprocessing, the chunks are just the size they are. It is not meaningful to "change the chunk size at train time".

> (with the caveat that for a large chunk size, some unfilled chunks might contain a ton of separator tokens)

You might be looking at the PackedDataset. The padding with separator tokens is a minor flaw of this dataset. But the `lightning.data` StreamingDataset/Preprocessor doesn't have this limitation. If a chunk doesn't get filled, it will just be made smaller. But that only happens at the tail of the dataset, as it is unlikely that your dataset is exactly divisible by the chunk size.

**Comment by jwkirchenbauer** [comment]
Right, I have been thinking about `PackedDataset`, since I am not operating on cloud storage, but rather on an attached disk. Do you still recommend the `StreamingDataset` even in this case?

On chunks, my bad, you're right, I suppose I should mean `block_size`. You seem to be able to load a `PackedDataset` with an arbitrary `block_size` at train time: [lit-gpt/pretrain/redpajama.py#L251-L252](https://github.com/Lightning-AI/lit-gpt/blob/main/pretrain/redpajama.py#L251-L252). I assume the fact that this isn't hardcoded into the header is to accommodate "prepare once, train at any sequence lengths" semantics.

I now see that the `n_chunks` train time parameter seems to be the prefetch size. So, that means the chunk size with which you write the packed dataset _will_ influence the min unit of prefetch you have available. Correct?

**Comment by awaelchli** [comment]
> Right, I have been thinking about PackedDataset, since I am not operating on cloud storage, but rather on an attached disk. Do you still recommend the StreamingDataset even in this case?

StreamingDataset will eventually replace PackedDataset, it's just not yet released officially. If you're curious, feel free to already use it  and feedback for it would be appreciated. 

Yes the block size / sequence length can be changed at training time, block size << chunk size.

**Comment by jwkirchenbauer** [comment]
Will do! Thanks for all the clarifications, and for the great work on the library
